llvm/lib/Target/AIE/AIEPostSelectOptimize.cpp

```diff
@@ -57,197 +57,139 @@ static cl::opt<bool>
                              cl::desc("Enable post select optimize."));
 
 namespace {
-class AIEPostSelectOptimize : public MachineFunctionPass {
-public:
-  static char ID;
-  AIEPostSelectOptimize();
-  StringRef getPassName() const override { return "AIE Post Select Optimizer"; }
-  bool runOnMachineFunction(MachineFunction &MF) override;
-  void getAnalysisUsage(AnalysisUsage &AU) const override;
-
-private:
-  const MachineRegisterInfo *MRI;
-
-  /**
-   * @struct PhysRegDefInfoStruct
-   * @brief To track the PhysReg defs and the state of their use.
-   */
-  struct PhysRegDefInfoStruct {
-    // Tracks the last visited def. Change with every new def.
-    MachineInstr *LastVisitedDef = nullptr;
-    // Tracks if the PhysReg is used in between current def and LastVisitedDef.
-    bool IsUsed = false;
-
-    PhysRegDefInfoStruct(MachineInstr *MI)
-        : LastVisitedDef(MI), IsUsed(false) {}
-  };
 
-  /// Collect all VirtReg to PhysReg copies that can be removed at the last.
-  DenseMap<Register, MachineInstr *> ErasableVirtRegToPhysRegCopies;
-
-  bool doPeepholeOpts(MachineBasicBlock &MBB);
-  bool eraseOrStoreRedundantPhysRegDefs(
-      MachineInstr &MI, DenseMap<Register, PhysRegDefInfoStruct> &PhysRegDefs);
+/// Information about a COPY that can be tracked by PhysRegCopyTracker.
+struct TrackableCopyOps {
+  Register VirtReg;
+  MCRegister PhysReg;
+  bool InvalidatePhysReg; // Whether PhysReg is being redefined.
+};
 
-  /// Checks if the \p Op is an unallocatable phys reg.
-  bool isUnallocatablePhysReg(const MachineOperand &Op);
+/// Track SSA copies of simple physical registers.
+/// Simple means they don't have sub or super registers.
+class PhysRegCopyTracker {
```

Comment:
I think this is more than a Tracker.

Reply:
Can't argue :D I can maybe make `trySimplifyCopy` a free function, then it would be just a tracker. What do you think?